### ISSUE TITLE
pg transition: org profile view

### ISSUE THREAD

**Issue by JMurphyWeb** [emfoundation/ce100-app#395]
Path: /orgs/{id}  [Handler](https://github.com/emfoundation/ce100-app/blob/master/server/handlers/org-details-view.js)
Current view:
![screen shot 2016-11-29 at 16 07 19](https://cloud.githubusercontent.com/assets/12638261/20717668/23db74fe-b64e-11e6-8a7c-71600e890882.png)
![screen shot 2016-11-29 at 16 07 37](https://cloud.githubusercontent.com/assets/12638261/20717672/26c59244-b64e-11e6-837e-531e66dd021f.png)

[**Public view in wireframes**](https://app.zeplin.io/project.html#pid=57ac9b5692d0c316520caeca&sid=57d985d9895d1e04118eff10)

[**Private view in wireframes**](https://app.zeplin.io/project.html#pid=57ac9b5692d0c316520caeca&sid=57d9907a3b1cfd1b11eb9d91)

[**Admin view in wireframes**](https://app.zeplin.io/project.html#pid=57ac9b5692d0c316520caeca&sid=57ceaaba78329bc82ced604b)

Data that is needed:
**org:**
name, 
id, 
mission statement, 
logo url, 
tags (name & id) - sorted alphabetically
**people:**
primary user attached to this organisation
name, 
phone number, 
job title, 
email address
**challenges:**
must be active and created by this organisation
each challenge title, description, id
tags associated with each challenge
suggested matches  on each challenge
-> need all organisations that meet a range of tags,
ordered by most tags matched, 
maximum of 10 matches per challenge

@asimonw please can you check over my interpretation of the permissions associated with this view:
Permissions:
ADMIN:
Can view active and inactive organisations
Edit icon shows on following cards: mission statement, tags
Edit icon does not show up on challenges (as admin cannot edit challenges)
Can see 'suggested matches' <- @asimonw do we still want this viewable on the org profile?
Link to see archived challenges

PRIMARY USER VIEWING OWN ORG:
Can view active organisations only
Edit icon shows on following cards: mission statement, tags, each challenge

PUBLIC VIEW (SECONDARY USER, OR PRIMARY VIEWING AN ORG THAT IS NOT THEIRS)
Can view active organisations only

**Comment by JMurphyWeb** [comment]
Possible query:
```sql
-- OUTER QUERY DEALS WITH ORGANISATION DETAILS, PRIMARY USER
-- ORGANISATION TAGS
SELECT 
organisations.id as org_id,
organisations.name as org_name,
people.id as primary_id,
people.first_name as primary_first_name,
tags.name as org_tag_name,
tags.id as org_tag_id,
chal.chal_title as challenge_title,
chal.tag_name as challenge_tag_name,
chal.tag_id as challenge_tag_id
FROM (
-- SUB QUERY GETS ALL ACTIVE CHALLENGES RELATED TO GIVEN ORG
  SELECT 
  challenges.title as chal_title,
  challenges.id as chal_id,
  tags.name as tag_name,
  tags.id as tag_id
  FROM challenges
  JOIN tags_challenges
  ON challenges.id = tags_challenges.challenges_id
  JOIN tags 
  ON tags_challenges.tags_id = tags.id
  WHERE challenges.org_id = 1
  AND tags.active = true
  AND challenges.active = true
  
) as chal
JOIN
organisations
ON
organisations.id = 1
AND organisations.active = true
JOIN
people
ON
organisations.id = people.org_id
AND people.user_type = 'primary'
AND people.active = true
JOIN
tags_organisations
ON
organisations.id = tags_organisations.organisations_id
JOIN
tags
ON
tags_organisations.tags_id = tags.id
;
-- WHEN WE JOIN THE INNER AND THE OUTER, WE GET 'REPEATED' RESULTS
-- BUT THE QUERY DOES SEEM TO WORK AS EXPECTED


```

**Comment by asimonw** [comment]
@JMurphyWeb Yes, look at the latest version of the wireframes. We can remove older versions if that's easier. 

As to the private and admin views: there seem to be small discrepancies between them (at least the suggested matches look different). @fannyBkane could you have a look at this and indicate which one we should use? There also seems to be an error on the admin view, as @JMurphyWeb mentioned admins shouldn't have an edit button on challenge cards. 

Originally, we also didn't want admins to be able to add a challenge and see archived challenges. So admin and private view are the same apart from:
* navigation
* edit button on challenge cards (not for admin)
* Add challenge (not for admin)
* See archived challenges (not for admin)

But to be honest, apart from the navigation, which definitely needs to be different, if it's technically easier to keep everything else exactly the same, that would be fine for now. Admins would then simply be able to do a bit more than is strictly necessary.

**Comment by fannyBkane** [comment]
Please use [Private view in wireframes](https://app.zeplin.io/project.html#pid=57ac9b5692d0c316520caeca&sid=57d9907a3b1cfd1b11eb9d91)

I have deleted the other one.

**Comment by JMurphyWeb** [comment]
Great thanks guys. I wanted to ask about suggested matches. I think I remember you saying at some point that you weren't sure about including suggested matches in this view as it can be quite a lot of scrolling..

Is that something that was mentioned at all?

**Comment by JMurphyWeb** [comment]
Now integrating our new function into the application.

**Comment by JMurphyWeb** [comment]
Integration went well so far. I needed to make an adjustment to the tags array. 

Still need to look into different permissions and add tests.

**Comment by JMurphyWeb** [comment]
Currently it fails for orgs with id of 2 or 3.
Probably down to no user being attached: https://github.com/postgres-plugin/people/issues/24 
need to check the mock data

**Comment by JMurphyWeb** [comment]
Exploring why this query is falling down for org id 2:

This query seems to do the job: we need to LEFT OUTER JOIN the last two joins:

```sql
SELECT
organisations.id as org_id, chal.chal_id as challenge_id, people.id as primary_id, chal.tag_id as challenge_tag_id
FROM (
  SELECT challenges.title as chal_title, challenges.id as chal_id, challenges.description as chal_description, tags.name as tag_name, tags.id as tag_id
  FROM challenges
  JOIN tags_challenges
  ON challenges.id = tags_challenges.challenges_id
  JOIN tags ON tags_challenges.tags_id = tags.id
  WHERE challenges.org_id = 2
  AND tags.active = true AND challenges.active = true
) as chal
JOIN organisations
ON organisations.id = 2 AND organisations.active = true
JOIN people ON organisations.id = people.org_id AND people.user_type = 'primary' AND people.active = true AND people.account_activated = true
LEFT OUTER JOIN tags_organisations ON tags_organisations.organisations_id = 2
LEFT OUTER JOIN tags ON tags_organisations.tags_id = tags.id ;

```

For reference, if we do not include the last two lines at all, we get the following table:

```
 org_id | challenge_id | primary_id | challenge_tag_id
--------+--------------+------------+------------------
      2 |            4 |          4 |                9
      2 |            4 |          4 |               11
      2 |            4 |          4 |               60
      2 |            4 |          4 |               69
```
with 4 entries, one for each tag

Then if we include the following line at the end:
```
JOIN tags_organisations ON tags_organisations.organisations_id = 2;
```
we get no results.

This is because we are trying to join on a table that does not meet the criteria: there are no rows in tags_organisations with organisation id of 2.  Therefore we need to left outer join to include all of our entries from all of the joins before this line.

**Comment by JMurphyWeb** [comment]
It also seems sensible to use LEFT OUTER JOIN for this join as well:

```
JOIN people ON organisations.id = people.org_id AND people.user_type = 'primary' AND people.active = true AND people.account_activated = true
```
which should help with this potential issue (no user on the org)

**Comment by JMurphyWeb** [comment]
```sql
SELECT
organisations.id as org_id, organisations.name as org_name, organisations.logo_url as org_logo_url, organisations.mission_statement as org_mission_statement,
people.id as primary_id, people.first_name as primary_first_name, people.last_name as primary_last_name, people.email as primary_email, people.phone as primary_phone, people.job_title as primary_job_title,
tags.name as org_tag_name, tags.id as org_tag_id,
chal.chal_title as challenge_title, chal.chal_description as challenge_description, chal.tag_name as challenge_tag_name, chal.tag_id as challenge_tag_id, chal.chal_id as challenge_id
FROM (
  SELECT challenges.title as chal_title, challenges.id as chal_id, challenges.description as chal_description, tags.name as tag_name, tags.id as tag_id
  FROM challenges
  JOIN tags_challenges
  ON challenges.id = tags_challenges.challenges_id
  JOIN tags ON tags_challenges.tags_id = tags.id
  WHERE challenges.org_id = 2
  AND tags.active = true AND challenges.active = true
) as chal
RIGHT OUTER JOIN organisations
ON organisations.active = true
LEFT OUTER JOIN people ON organisations.id = people.org_id AND people.user_type = 'primary' AND people.active = true AND people.account_activated = true
LEFT OUTER JOIN tags_organisations ON tags_organisations.organisations_id = 2
LEFT OUTER JOIN tags ON tags_organisations.tags_id = tags.id
WHERE organisations.id = 2;
```

**Comment by JMurphyWeb** [comment]
Ended up settling on this.

**Comment by JMurphyWeb** [comment]
Seems to be a problem with my formatting function. Going to add some tests to the plugin.

**Comment by JMurphyWeb** [comment]
Above issue was due to the formatter adding a default entry into the tags array.

I have changed it so that it will return an empty array if there are no tags.

**Comment by JMurphyWeb** [comment]
I have noticed another issue with it: if there are no tags in a given challenge, that challenge is not returned in the query.  Will look into more outer joins.

**Comment by JMurphyWeb** [comment]
So the inner query did not include any challenges that did not have tags due to this line:
```sql
  JOIN tags ON tags_challenges.tags_id = tags.id
```
Amended the inner query to something like this:
```sql
SELECT 
* 
FROM CHALLENGES 
LEFT OUTER JOIN tags_challenges ON challenges.id = tags_challenges.challenges_id 
LEFT OUTER JOIN tags on tags_challenges.tags_id = tags.id 
where challenges.org_id = 2;
```

**Comment by JMurphyWeb** [comment]
Inner query was actually still failing when I only checked for active tags.

```sql
  SELECT challenges.title AS chal_title, challenges.id AS chal_id, challenges.description AS chal_description, tags.name AS tag_name, tags.id AS tag_id, tags.active AS tag_active, challenges.active AS chal_active
  FROM CHALLENGES
  LEFT OUTER JOIN tags_challenges ON challenges.id = tags_challenges.challenges_id
  LEFT OUTER JOIN tags ON tags_challenges.tags_id = tags.id
  WHERE challenges.org_id = 2
  -- IF A CHALLENGE HAS NO TAGS, ITS TAGS.ACTIVE WILL BE NULL
  AND (tags.active IS NOT FALSE)
  AND challenges.active = true
```

**Comment by SimonLab** [comment]
resolved with #403
